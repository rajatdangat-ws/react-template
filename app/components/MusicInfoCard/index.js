/**
 *
 * MusicInfoCard
 *
 */

import React, { memo, useState, useEffect, useRef } from 'react';
import { useHistory } from 'react-router-dom';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { Card } from 'antd';
import { PlayCircleOutlined, PauseCircleOutlined, InfoCircleOutlined } from '@ant-design/icons';

const { Meta } = Card;

const StyledMeta = styled(Meta)`
  & .ant-card-meta-description {
    white-space: nowrap;
    overflow-y: hidden;
    text-overflow: ellipsis;
  }
`;

function MusicInfoCard({ trackName, coverImgUrl, artistName, previewUrl, detailsUrl, onActionButtonClick }) {
  const history = useHistory();
  const audioRef = useRef();
  const isMountedRef = useRef(false);

  const [isPlaying, setIsPlaying] = useState(false);

  useEffect(() => {
    isMountedRef.current = true;

    audioRef.current.onended = () => {
      setIsPlaying(false);
    };
    audioRef.current.onpause = () => {
      if (isMountedRef.current) {
        setIsPlaying(false);
      }
    };
    audioRef.current.onplay = () => {
      setIsPlaying(true);
    };

    return () => {
      isMountedRef.current = false;
    };
  }, []);

  const handleMoreClick = () => {
    history.push(detailsUrl);
  };

  const handlePlayClick = () => {
    onActionButtonClick(audioRef.current);
    if (isPlaying) {
      audioRef.current.pause();
    } else {
      audioRef.current.play();
    }
  };

  return (
    <>
      <Card
        data-testid="music-info-card"
        cover={<img src={coverImgUrl} alt={trackName} height="250px" width="250px" style={{ objectFit: 'cover' }} />}
        actions={[
          isPlaying ? (
            <PauseCircleOutlined data-testid="pause-button" key="pause" onClick={handlePlayClick} />
          ) : (
            <PlayCircleOutlined data-testid="play-button" key="play" onClick={handlePlayClick} />
          ),
          <InfoCircleOutlined data-testid="more-button" key="info" onClick={handleMoreClick} />
        ]}
      >
        <StyledMeta title={trackName} description={artistName} />
      </Card>
      <audio data-testid="audio-element" ref={audioRef} src={previewUrl} />
    </>
  );
}

MusicInfoCard.propTypes = {
  trackName: PropTypes.string,
  coverImgUrl: PropTypes.string,
  artistName: PropTypes.string,
  previewUrl: PropTypes.string,
  detailsUrl: PropTypes.string,
  onActionButtonClick: PropTypes.func
};

export default memo(MusicInfoCard);
