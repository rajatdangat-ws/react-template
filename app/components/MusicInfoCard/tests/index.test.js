/**
 *
 * Tests for MusicInfoCard
 *
 */

import React from 'react';
// import { fireEvent } from '@testing-library/dom'
import userEvent from '@testing-library/user-event';
import { renderWithIntl, timeout } from '@utils/testUtils';
import MusicInfoCard from '../index';

const props = {
  trackName: 'test track',
  coverImgUrl: 'image url',
  artistName: 'test artist',
  previewUrl:
    'https://audio-ssl.itunes.apple.com/itunes-assets/AudioPreview125/v4/ed/c5/2c/edc52c92-d0bd-d9ea-e1b2-25b89c90e3ac/mzaf_2299253512947720106.plus.aac.p.m4a',
  detailsUrl: 'details url'
};

jest.mock('react-router-dom');

describe('<MusicInfoCard />', () => {
  it('should render and match the snapshot', () => {
    const { baseElement } = renderWithIntl(<MusicInfoCard />);
    expect(baseElement).toMatchSnapshot();
  });

  it('should contain 1 MusicInfoCard component', () => {
    const { getAllByTestId } = renderWithIntl(<MusicInfoCard {...props} />);
    expect(getAllByTestId('music-info-card').length).toBe(1);
  });

  it('should display music title and artist', () => {
    const { getByText } = renderWithIntl(<MusicInfoCard {...props} />);
    expect(getByText('test track')).not.toBeNull();
  });

  it('should go to music info page when clicked on more button', async () => {
    const routerDom = require('react-router-dom');
    const pushSpy = jest.fn();
    const history = jest.spyOn(routerDom, 'useHistory').mockImplementation(() => {
      return {
        push: pushSpy
      };
    });

    const { getByTestId } = renderWithIntl(<MusicInfoCard {...props} />);
    const button = getByTestId('more-button');
    const user = userEvent.setup();
    expect(button).not.toBeNull();
    await user.click(button);

    expect(history).toBeCalled();
    expect(pushSpy).toBeCalledWith('details url');
  });

  it('should contain audio tag with correct url', () => {
    const { getByTestId } = renderWithIntl(<MusicInfoCard {...props} />);

    const audioEl = getByTestId('audio-element');
    expect(audioEl).toHaveAttribute('src', props.previewUrl);
  });

  it('should play and pause music when clicked on play and pause buttons', async () => {
    const { getByTestId } = renderWithIntl(<MusicInfoCard {...props} onActionButtonClick={() => {}} />);
    const user = userEvent.setup();
    const audioElement = getByTestId('audio-element');
    await timeout(500);

    let playBtn = getByTestId('play-button');
    expect(playBtn).toBeInTheDocument();

    await user.click(playBtn);
    expect(playBtn).not.toBeInTheDocument();
    expect(audioElement.paused).toBe(false);

    await timeout(500);

    const pauseBtn = getByTestId('pause-button');
    expect(pauseBtn).toBeInTheDocument();

    await user.click(pauseBtn);
    expect(audioElement.paused).toBe(true);
    expect(pauseBtn).not.toBeInTheDocument();
    playBtn = getByTestId('play-button');
    expect(playBtn).toBeInTheDocument();
  });

  it('should update the UI when the music finishes playing', async () => {
    const { getByTestId } = renderWithIntl(<MusicInfoCard {...props} onActionButtonClick={() => {}} />);
    const user = userEvent.setup();

    let playBtn = getByTestId('play-button');
    const audioElement = getByTestId('audio-element');

    expect(playBtn).toBeInTheDocument();
    await user.click(playBtn);
    const pauseBtn = getByTestId('pause-button');
    expect(pauseBtn).toBeInTheDocument();
    await timeout(500);
    audioElement.emulateStop();
    playBtn = getByTestId('play-button');
    expect(playBtn).toBeInTheDocument();
  });
});
