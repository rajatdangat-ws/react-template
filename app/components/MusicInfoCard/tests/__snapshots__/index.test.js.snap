// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MusicInfoCard /> should render and match the snapshot 1`] = `
<body>
  <div>
    <div
      class="ant-card ant-card-bordered MusicInfoCard__InfoCard-sc-4tgp7y-1 hXHvIE"
      data-testid="music-info-card"
    >
      <div
        class="ant-card-cover"
      >
        <img
          height="250px"
          loading="lazy"
          style="max-inline-size: 100%;"
          width="250px"
        />
      </div>
      <div
        class="ant-card-body"
      >
        <div
          class="ant-card-meta MusicInfoCard__StyledMeta-sc-4tgp7y-0 kOaJye"
        />
      </div>
      <ul
        class="ant-card-actions"
      >
        <li
          style="width: 50%;"
        >
          <span>
            <span
              aria-label="play-circle"
              class="anticon anticon-play-circle"
              data-testid="play-button"
              role="img"
              tabindex="-1"
            >
              <svg
                aria-hidden="true"
                data-icon="play-circle"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="64 64 896 896"
                width="1em"
              >
                <path
                  d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z"
                />
                <path
                  d="M719.4 499.1l-296.1-215A15.9 15.9 0 00398 297v430c0 13.1 14.8 20.5 25.3 12.9l296.1-215a15.9 15.9 0 000-25.8zm-257.6 134V390.9L628.5 512 461.8 633.1z"
                />
              </svg>
            </span>
          </span>
        </li>
        <li
          style="width: 50%;"
        >
          <span>
            <span
              aria-label="info-circle"
              class="anticon anticon-info-circle"
              data-testid="more-button"
              role="img"
              tabindex="-1"
            >
              <svg
                aria-hidden="true"
                data-icon="info-circle"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="64 64 896 896"
                width="1em"
              >
                <path
                  d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z"
                />
                <path
                  d="M464 336a48 48 0 1096 0 48 48 0 10-96 0zm72 112h-48c-4.4 0-8 3.6-8 8v272c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8V456c0-4.4-3.6-8-8-8z"
                />
              </svg>
            </span>
          </span>
        </li>
      </ul>
    </div>
    <audio
      data-testid="audio-element"
    />
  </div>
</body>
`;
